src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LogisticRegression.cs

```diff
@@ -397,4 +426,113 @@ public static CommonOutputs.BinaryClassificationOutput TrainBinary(IHostEnvironm
+        /// Due to the existence of regularization, an approximation is used to compute the variances of the trained linear coefficients.
+        /// </summary>
+        VBuffer<float> ComputeStd(double[] hessian, int[] weightIndices, int parametersCount, int currentWeightsCount, IChannel ch, float l2Weight);
+    }
+
+    public sealed class ComputeLRTrainingStd: IComputeLRTrainingStd
+    {
+        /// <summary>
+        /// Computes the standart deviation matrix of each of the non-zero training weights, needed to calculate further the standart deviation,
+        /// p-value and z-Score.
+        /// If you need faster calculations, use the ComputeStd method from the Microsoft.ML.HALLearners package, which makes use of hardware acceleration.
+        /// Due to the existence of regularization, an approximation is used to compute the variances of the trained linear coefficients.
+        /// </summary>
+        /// <param name="hessian"></param>
+        /// <param name="weightIndices"></param>
+        /// <param name="numSelectedParams"></param>
+        /// <param name="currentWeightsCount"></param>
+        /// <param name="ch">The <see cref="IChannel"/> used for messaging.</param>
+        /// <param name="l2Weight">The L2Weight used for training. (Supply the same one that got used during training.)</param>
+        public VBuffer<float> ComputeStd(double[] hessian, int[] weightIndices, int numSelectedParams, int currentWeightsCount, IChannel ch, float l2Weight)
+        {
+            Contracts.AssertValue(ch);
+            Contracts.AssertValue(hessian, nameof(hessian));
+            Contracts.Assert(numSelectedParams > 0);
+            Contracts.Assert(currentWeightsCount > 0);
+            Contracts.Assert(l2Weight > 0);
+
+            double[,] matrixHessian = new double[numSelectedParams, numSelectedParams];
+
+            int hessianLength = 0;
+            int dimention = numSelectedParams - 1;
+
+            for (int row = dimention; row >= 0; row--)
+            {
+                for (int col = 0; col <= dimention; col++)
+                {
+                    if ((row + col) <= dimention)
+                    {
+                        if ((row + col) == dimention)
+                        {
+                            matrixHessian[row, col] = hessian[hessianLength];
+                        }
+                        else
+                        {
+                            matrixHessian[row, col] = hessian[hessianLength];
+                            matrixHessian[dimention - col, dimention - row] = hessian[hessianLength];
+                        }
+                        hessianLength++;
+                    }
+                    else
+                        continue;
+                }
+            }
+
+            var h = Matrix<double>.Build.DenseOfArray(matrixHessian);
```

Comment:
I must be missing something here, so please forgive my ignorance.

What exactly are we using Math.NET for in this code? Are we just using it for the Matrix type, and for Inversing a Matrix? Or is there something else we are using that I'm not seeing?

If it is just for the Matrix type and Inverse, I don't think a full dependency on Math.NET is necessary.

Reply:
Not missing anything. Math.Net numeric is used just to calculate the inverse. We discussed the option to  implement it ourselves, and it was deemed error prone to round-of errors and unnecessary to own that code. 
There is also a parallel pending discussion on weather we need  to have an alternative method to MKL, which seems like it is leaning towards just keeping the MKL version. 

---
In reply to: [232706971](https://github.com/dotnet/machinelearning/pull/1392#discussion_r232706971) [](ancestors = 232706971)

Reply:
OK, thanks for the information.